### ISSUE TITLE
Does the multipart tokenizer really need to copy so many strings?

### ISSUE THREAD

**Issue by Yoric** [binast/binjs-ref#81]
We may wish to replace our use of `Vec<u8>` with something closer to a Rope.

**Comment by Yoric** [comment]
We should then do two passes:

1. Rewrite the tree to replace `Offset` with a byte length, also updating statistics.
2. Actually write data.

**Comment by Yoric** [comment]
Actually, we have several problems that we need to solve:

- When the API client writes a string, this string is actually put in a table. Instead, we put a placeholder `String(index)` marker. The `index` is only resolved after we have added all strings, we wish to ensure that most common strings have the smallest `index`.
- Similarly, when the API client writes a node, the node is actually put in a table. Instead, we put a placeholder marker. The index, once again, is only resolved after we have added all nodes.
- We have lists, but the 
- When the API client writes an Offset, the offset is replaced by a placeholder. The Offset is resolved after we have finished writing the rest of the contents of the node.

So, we could have the following structure:

```rust
/// The tree we build as we go.
enum DataAsItIsBuilt {
   UnresolvedStringIndex(?),
   UnresolvedTableIndex(?),
   UnresolvedOffset(Box<DataAsItIsBuilt>),
   Binary(Vec<u8>),
   Tuple(Vec<DataAsItIsBuilt>)
}

/// The tree we're going to write.
enum ResolvedData {
   Binary(Vec<u8>),
   Tuple(Vec<ResolvedData>),
}

impl DataAsItIsBuilt {
  fn resolve(self) -> ResolvedData;
}
```

This should make the code more readable and remove most copies. We may lose some statistics, though.
